pkg/operators/v1alpha1/catalogsource_types.go

```diff
@@ -96,7 +100,30 @@ type RegistryPoll struct {
 	// Interval is used to determine the time interval between checks of the latest catalog source version.
 	// The catalog operator polls to see if a new version of the catalog source is available.
 	// If available, the latest image is pulled and gRPC traffic is directed to the latest catalog source.
-	Interval *metav1.Duration `json:"interval,omitempty"`
+	Interval     *metav1.Duration `json:"interval,omitempty"`
+	ParsingError string           `json:"-"`
+}
+
+// UnmarshalJSON implements the encoding/json.Unmarshaler interface.
+func (u *UpdateStrategy) UnmarshalJSON(data []byte) (err error) {
+	var updateStrategyMap map[string]map[string]string
+	if err = json.Unmarshal(data, &updateStrategyMap); err != nil {
+		return err
+	}
+	registryPoll := &RegistryPoll{}
+	duration, err := time.ParseDuration(updateStrategyMap["registryPoll"]["interval"])
+	if err != nil {
+		registryPoll.ParsingError = fmt.Sprintf("error parsing spec.updateStrategy.registryPoll.interval. Setting the default value of %v. Error:%v", DefaultRegistryPollDuration, err)
+		defaultTime, parseErr := time.ParseDuration(DefaultRegistryPollDuration)
```

Comment:
I think we will immediately run into problems if we overwrite the given value with a default.

Let's walk through how:
1. User creates a CatalogSource with an invalid interval
2. Informers detect the change and attempt unmarshalling
3. The custom unmarshaller detects the invalid interval, sets the parsing error and sets the default interval
4. Custom unmarshalled CatalogSource enter our sync handlers
5. Sync handlers see the parsing error and set the invalid interval status condition
6. Sync handlers update CatalogSource with:
  - __valid__ default interval
  - invalid interval status condition
7. Informers detect the change and attempt unmarshalling
8. __The custom unmarshaller sees a valid interval, doesn't set parseError__
9. Custom unmarshalled CatalogSource enter our sync handlers
10. __Sync handlers see no parsing error and drop the invalid interval parsing condition__
11. If the user blinked, they would have no idea that:
  -  the CatalogSource they just created had an invalid status
  -  the CatalogSource has a different interval than what they intended

Reply:
That is a great catch @njhale . I've updated the struct to include a `RawInterval` field that'll contain the user input, while the current `Interval` field will hold the parsed duration. Caveat is that if the default is set, the user will still see the ill format-ed input in the catsrc spec, but the status condition will contain a message stating that the default is being used instead. This looks like our best option in terms of what's least surprising for our users.

Reply:
Is there any reason to set a default at all? Couldn't we leave the catalog in an inoperable state until it's fixed?

Reply:
@anik120 This is likely an artifact of too many context switches to apologies if you've already answered this: can we simply just bubble up the registryPoll.ParsingError without setting a default? I was playing around with this locally by manually updating the o-f/api vendored code in my OLM fork and was surprised by the behavior (although it did fix the problematic behavior of failing to establish a watch).

Reply:
> Is there any reason to set a default at all? Couldn't we leave the catalog in an inoperable state until it's fixed?

My memory is a little hazy around this too. On the surface, I'm with @perdasilva and @timflannagan here, unless @anik120 could provide some reasoning for the default in the invalid case.

Reply:
@perdasilva @njhale @timflannagan so getting an error in unmarshalling the value is what started this fix in the first place. If there is no custom unmarshaller, a value like "15mError.code" gets the controller stuck in a loop trying to set up a watch for the catsrc. If we return an error from the custom unmarshaller, it takes us back to the same problem, the controller gets stuck in a loop trying to set up a watch for the catsrc. Since the correct way to avoid this is schema validation, but that is not back portable in this case, handling this error in this way and setting a default is the only correct solution for this problem.

Reply:
Also, since we get into the same problem of failing to set up a watch for the catsrc, none of the syncs are performed for the catsrc, so the catalog operator logic of checking for the `parsingError` and setting the status condition if there is an error is never executed. i.e the users are left in the dark about why the catsrc they added did not do anything for them, which forces them to inspect the pod logs, which is another thing we are trying to avoid.

Reply:
Here's the PR: https://github.com/operator-framework/operator-lifecycle-manager/pull/2447